### ISSUE TITLE
[Local GC] What to do with Volatile<T> and Interlocked

### ISSUE THREAD

**Issue by swgillespie** [dotnet/coreclr#13569]
The [`Volatile<T>` class](https://github.com/dotnet/coreclr/blob/master/src/inc/volatile.h#L251) is designed to reconcile the major differences between `volatile`-qualified types when compiled by MSVC and clang/gcc. In particular, stores to volatile objects have release semantics and reads from volatile objects have acquire semantics (without `/volatile:iso`, which we don't use). The C++ standard does not guarantees this and thus clang/gcc provide no such guarantees when writing or reading from volatile objects.

This class is used pervasively throughout the runtime and the GC. However, the GC (in general) refers to it through the `VOLATILE(T)` macro, which expands to `T volatile` when building standalone or when building as a part of CoreRT. This is incorrect by C++ standards and, while it's not going to cause any issues when compiled with MSVC volatile semantics, it will break in unexpected ways when compiled with gcc/clang.

Therefore, it seems that we're going to need some sort of `Volatile<T>` replacement that can be used by a standalone/CoreRT GC. Yesterday I provided an implementation of `Volatile<T>` using `std::atomic<T>`, which works for 95% of the GC's uses of `Volatile<T>` (see implementation: https://github.com/swgillespie/coreclr/blob/local-gc-volatile/src/gc/env/volatile.h#L26). The other 5% of the GC's uses of `Volatile<T>` involve the GC using `Volatile<T>` fields with methods on the `Interlocked` class. In order to make this work, `Volatile<T>` provides a conversion operator to `T volatile*`, which is hugely unsafe and can't be done using `std::atomic<T>` for that reason.

It seems to me that all functions on `Interlocked` can be implemented as methods on `Volatile<T>`. This is especially true if we implement `Volatile<T>` using `std::atomic<T>`, in which case the implementation is trivial (forwarding trivially to methods on `std::atomic<T>`). Furthermore, looking at the code, the GC seems to be the only component that is using the `Interlocked` class itself (most of the runtime seems to forward to the Win32/PAL Interlocked* methods instead).

I think that this could be a good opportunity to eliminate the `Interlocked` class and move its functionality to `Volatile<T>` instead. This wold involve adding at a minimum `Exchange` and `CompareExchange` methods on `Volatile<T>`. The runtime's implementation of `Volatile<T>` will likely need to keep its conversion operator to `T volatile*` (due to usage in the PAL) but the GC's implementation of `Volatile<T>` (in terms of `std::atomic`) will not.

**Comment by swgillespie** [comment]
cc @jkotas @janvorli @Maoni0 @sergiy-k - Any thoughts on this? Now that a standalone GC builds without VM headers, this is going to be a problem going forward I think.

**Comment by jkotas** [comment]
> This is incorrect by C++ standards

Why is this incorrect?

> Yesterday I provided an implementation of Volatile<T> using std::atomic<T>

This is good. Any reason why the GC cannot just call `std::atomic<T>` directly? What does `Volatile<T>` abstraction gives us?

**Comment by jkotas** [comment]
> without /volatile:iso, which we don't use

We should be using it. We have used it in the old razzle build. I guess it got missed during conversion to CMake.

**Comment by swgillespie** [comment]
From http://en.cppreference.com/w/cpp/language/cv:

> Every access (read or write operation, member function call, etc.) made through a glvalue expression of volatile-qualified type is treated as a visible side-effect for the purposes of optimization (that is, within a single thread of execution, volatile accesses cannot be optimized out or reordered with another visible side effect that is sequenced-before or sequenced-after the volatile access. This makes volatile objects suitable for communication with a signal handler, but not with another thread of execution, see std::memory_order).

**Comment by swgillespie** [comment]
> Any reason why the GC cannot just call std::atomic<T> directly? What does Volatile<T> abstraction gives us?

My understanding is that, when not building standalone, the GC can't use system headers at all, even c++ standard ones. I just tried to include `<atomic>` in gc.cpp and got a lot of errors, so it seems to me that this is the case.

**Comment by swgillespie** [comment]
also cc @sdmaclea, who may be interested

**Comment by jkotas** [comment]
>  when not building standalone, the GC can't use system headers at all

To solve this problem, I would add our own `atomic` that wraps the PAL, etc. but has same methods as the standard atomic.

**Comment by jkotas** [comment]
> From http://en.cppreference.com/w/cpp/language/cv:

This does not make it incorrect. It depends on how you write the code. When we have introduced `Volatile<T>`, Patrick insisted on not using it in the GC because of he wanted to have the barriers always explicit (e.g. by calling Interlocked methods). It is why the `VOLATILE` macro exists.

**Comment by swgillespie** [comment]
Ah, makes sense! I can understand wanting the barriers to be explicit.

Would this be an appropriate place for such an `atomic` wrapper? https://github.com/dotnet/coreclr/tree/master/src/inc/clr_std

**Comment by jkotas** [comment]
Yes.

**Comment by janvorli** [comment]
Regarding the /volatile:iso, for ARM, it is the default that the compiler uses. For intel, it uses /volatile:ms by default, but the /volatile:ms and /volatile:iso are equivalent on intel. So not specifying it explicitly doesn't hurt.

**Comment by Maoni0** [comment]
I am a bit concerned with possible perf degrade. It seems like we almost need 2 classes - one is for just volatile and the other for "atomic volatile" which is what you implemented. volatile and atomic are orthogonal. It's just that often when we use a volatile we also want the atomic semantic - not always, for example the [volatile](https://github.com/dotnet/coreclr/issues/13015) we had to add to get around the compiler optimization so it doesn't replace a pattern with memset unwantedly. For that one we definitely don't want to replace add with lock add. But you could argue that that's just a temporary solution and we'll replace it with something better that doesn't use volatile. But still we use volatile because we don't want the compiler to cache values and/or we want to enforce ordering. I played around a bit with the atomic stuff - for adding/substracting we almost always want the atomicity so using atomic is the same perf as interlocked. For reading/storing I didn't detect a perf difference between using store with memory_order_release/load with memory_order_acquire and volatile load/store. But I only did something very simple. Certainly we want to make sure we don't introduce perf degrade in perf critical paths.

**Comment by Maoni0** [comment]
Of course I just remembered that for the avoiding-memset case we just needed to make the location that gets set volatile so that would mean a store, not an add...

**Comment by swgillespie** [comment]
> It seems like we almost need 2 classes - one is for just volatile and the other for "atomic volatile" which is what you implemented.

Yeah, I agree. When do we need `volatile` at all, though? The optimization-inhibiting volatile is sufficient enough for us, but it seems to me that the most correct thing for us to do from a standards perspective would be to replace the optimization-inhibiting volatile with writes with `memory_order_relaxed` semantics, since anything else is undefined behavior.

I did some experiments (https://godbolt.org/g/DkjRbe) comparing the generated code between `std::atomic<T>` and `T volatile` + `Interlocked`. (Caveat - this is GCC, not clang, but I imagine that they are similar in the optimizations done related to the standard, like these are.) On Intel, ARM, and AArch64, the code for `StoreWithoutBarrier` is identical whether or not we use a volatile write or an atomic write with `memory_order_relaxed` semantics. I also see that the compiler inserts a barrier on ARM for a store with release semantics (but doesn't on AArch64?) Also the code for `Interlocked::CompareExchange` is worse...

**Comment by sdmaclea** [comment]
> but doesn't on AArch64

It should be using the store release assembly instruction `stl*`.  The `l` means release.

**Comment by janvorli** [comment]
> Also the code for Interlocked::CompareExchange is worse

That is true only if you don't let it get inlined. When inlined, the call to `Interlocked::CompareExchange((int volatile*)lock_, 2, 3);` becomes
```asm
        mov     r2, #2
        mov     r1, #3
        ldr     r0, .L5
        bl      __sync_val_compare_and_swap_4
```
while the `lock2_.CompareExchange(2, 3);` becomes
```asm
        mov     r3, #3
        mov     r2, #2
        mov     r1, r3
        ldr     r0, .L4
        str     r3, [sp, #4]
        bl      __sync_val_compare_and_swap_4
```

**Comment by swgillespie** [comment]
@sdmaclea ah, that's great then!

@janvorli I see, thanks! The stack reference in `lock2_.CompareExchange(2, 3)` looks unfortunate, though.

At any rate, would it be an acceptable temporary workaround for all of you to copy the `volatile.h` header into `src\gc\env` for now? The standalone GC build is currently broken due to the missing `Volatile<T>` type and it looks like we're going to need it or an `std::atomic` replacement one way or another.

**Comment by janvorli** [comment]
@swgillespie actually, I've failed to include the full  `lock2_.CompareExchange(2, 3)`. You can see it below and it explains why it has the stack reference. It is caused by the signature of the compare_exchange_weak function where the first parameter, the "expected", is a reference and it is set by the operation in case the target was not updated. 
```asm
        mov     r3, #3
        mov     r2, #2
        mov     r1, r3
        ldr     r0, .L10+4
        str     r3, [sp, #4]
        bl      __sync_val_compare_and_swap_4
        cmp     r0, #3
        strne   r0, [sp, #4]
```

**Comment by sdmaclea** [comment]
>  copy the volatile.h header into src\gc\env for now

@swgillespie I thought there were two copies of volatile.h already.  I am certainly OK with two copies.

**Comment by Maoni0** [comment]
I've spent some time looking into our volatile usage and std::atomics. I was concerned whether we have the freedom to do the non atomic stuff if we don't need to (eg, obviously don't want a++ to turn into an InterlockedInc if I don't need it to be). But we could just do 

```temp = a.load(memory_order_relaxed);```
```temp++;```
```a.store(temp, memory_order_relaxed);```

then we are not penalized (compared to what we have today in the Volatile implementation, that is). We don't have many places that do this but I wanted to make sure there is a way to do it for the rare occasions we need it.

I think our plan should be -

step 1) we change the code in a straightforward way, basically what you did. Load->load with memory_order_acq and Store->store with memory_order_rel and WithoutBarrier -> memory_order_relaxed. We can do some stress runs to make sure things look ok and I would expect no perf regression here as - we are the same on x86/x64 as the vanilla acq/rel are just normal load and stores (I haven't come across a store that needs to be SC in which case it would be a lot more expensive 'cause it needs to lock); on other platform/compiler combinations I would think using atomics might give us some perf gain if it avoids full barrier in cases that we are doing full barriers right now.

step 2) we see if there's any perf gain to be had - I know there is just by looking at a couple of places where we heavily use Volatile right now. This obviously will require another round of stress.
